router/dataplane.go

```diff
@@ -822,20 +829,27 @@ func (d *DataPlane) Run(ctx context.Context /*@, ghost place io.Place, ghost sta
 			// @ 	msgs[i].Mem() &&
 			// @ 	msgs[i].HasActiveAddr() &&
 			// @ 	msgs[i].GetAddr() == nil
+			// @ ensures forall j int :: { &msgs[j] } (0 <= j && j < len(msgs)) ==>
+			// @ 	sl.Bytes(msgs[j].GetFstBuffer(), 0 , len(msgs[j].GetFstBuffer()))
 			// @ decreases
 			// @ outline(
 			// @ invariant 0 <= i0 && i0 <= len(msgs)
 			// @ invariant forall i int :: { &msgs[i] } i0 <= i && i < len(msgs) ==>
 			// @ 	msgs[i].Mem() && msgs[i].GetAddr() == nil
 			// @ invariant forall i int :: { &msgs[i] } 0 <= i && i < i0 ==>
 			// @ 	msgs[i].Mem() && msgs[i].GetAddr() == nil && msgs[i].HasActiveAddr()
+			// @ invariant forall j, k int :: { &msgs[j], &msgs[k] } (0 <= j && j < k && k < i0) ==>
```

Comment:
```suggestion
			// @ invariant forall j, k int :: { &msgs[j], &msgs[k] } 0 <= j && j < k && k < i0 ==>
```

Reply:
It might be worth introducing a pure function for this if this expression shows up more often

Reply:
How would you do because if a write a function like this: 
```
ghost
requires forall i int :: { &msgs[i] } 0 <= i && i < len(msgs) ==> acc(msgs[i].Mem(), R55)
decreases
pure func MessagesUBufBytesPerms(msgs []Message) bool {
	return forall i int :: { &msgs[i] } 0 <= i && i < len(msgs) ==>
		sl.Bytes(msgs[i].GetFstBuffer(), 0, len(msgs[i].GetFstBuffer()))
}
```

I get the following error: `property error: assertion is not assignable to bool`

Reply:
Ah, that is not the line I selected. I selected this: `// @ invariant forall j, k int :: { &msgs[j], &msgs[k] } 0 <= j && j < k && k < i0 ==> msgs[j].GetFstBuffer() !== msgs[k].GetFstBuffer()`

Reply:
Ah, my bad. We only use it once. So I think a separate function is a bit overkill.